ringpop.thrift-gen

```diff
@@ -0,0 +1,221 @@
+{{ define "arguments" }}ctx {{ contextType }}{{ range .Arguments }}, {{ .Name }} {{ .Type | goType }}{{ end }}{{ end}}
```

Comment:
Is it necessary to have some of this intricate logic in generated code? Most of the implementation of the functions `getDestinations()`, `getClients()`, `getClientForDest()` do not require generation/templating whatsoever and are probably best served by baking them into core Ringpop APIs. It seems that only reason they are templated is because of their return type. Instead, can Ringpop preserve a cache of remote clients of `interface{}` and when it needs to generate a remote client, farm that off to Thrift-generated code?

For example, in core Ringpop, I'm thinking of a new API like, `NewRouter()`:

```
type RouterConfig struct {
    N ReplicationFactor
    Fanout ReplicationFactor
}

type RoutingClientGenerator interface {
    CreateClient() interface{}
}

func NewRouter(cfg RouterConfig, clientGenerator RoutingClientGenerator) (*RingpopRouter) {

}

func (*RingpopRouter) GetClients() ([]interface{}) {

}
```

then, in generated

```
clients := router.GetClients(...)
// cast to T
```